### ISSUE TITLE
[Bug] l2_bypass lowers to TLOAD<TLoadL2Hint::NotAllocKeep>, whose runtime offset nobody fills — silent wrong data or AICore fault on a2a3

### ISSUE THREAD

**Issue by zhangqi-chen** [hw-native-sys/PTOAS#1537]
## Background

In pypto-lib, the DeepSeek-V4 Flash routed-expert kernel
(`models/deepseek_v4_flash_mtp/expert_routed.py`) declares
`pl.set_cache_policy(w, pl.CachePolicy.BYPASS)` on its INT8 grouped-matmul
weights (hw-native-sys/pypto-lib#1039). On a2a3 that now produces a **precision /
runtime** failure: the bypassed weight load either returns garbage (golden FAIL,
no error reported) or raises an AICore exception that stalls the scheduler.

Reproduce with (real device required; a2a3sim does not reproduce):

    L2G_B1=0 L2G_B3=0 L2G_B2=1 python repro_l2_aic3.py --device N   # silent wrong data
    L2G_B1=1 L2G_B3=0 L2G_B2=0 python repro_l2_aic3.py --device N   # AICore exception
    L2G_B1=0 L2G_B3=0 L2G_B2=0 python repro_l2_aic3.py --device N   # control: exact

The repro is **not committed**. Save the file below and run it; it imports only
`pypto` and `torch`.

Reproduction case, `repro_l2_aic3.py`:

```python
"""Three AIC matmul kernels; each weight's L2 bypass is switched independently.

Mirrors expert_routed's gate / up / w2 kernel count.

    L2G_B1=0 L2G_B3=0 L2G_B2=1 python repro_l2_aic3.py --device N   # only the last kernel bypasses
"""
import argparse
import os

import torch

import pypto.language as pl
from pypto.runtime import RunConfig

E = int(os.environ.get("L2G_E", "4"))
MGI = int(os.environ.get("L2G_MGI", "4"))
R = 64
D = 4096
N = 2048
ROW_TILE = 16
N_TILE = 256
K_TILE = 512


def _cache(flag):
    return pl.CachePolicy.BYPASS if os.environ.get(flag, "0") == "1" else pl.CachePolicy.DEFAULT


C1 = _cache("L2G_B1")
C3 = _cache("L2G_B3")
C2 = _cache("L2G_B2")


@pl.jit
def three_mm(
    recv_x: pl.Tensor[[E, R, D], pl.INT8],
    recv_count: pl.Tensor[[E, 1], pl.INT32],
    w1: pl.Tensor[[E, N, D], pl.INT8],
    w3: pl.Tensor[[E, N, D], pl.INT8],
    w2: pl.Tensor[[E, N, D], pl.INT8],
    o1: pl.Out[pl.Tensor[[E * R, N], pl.INT32]],
    o3: pl.Out[pl.Tensor[[E * R, N], pl.INT32]],
    o2: pl.Out[pl.Tensor[[E * R, N], pl.INT32]],
):
    x_flat = pl.reshape(recv_x, [E * R, D])
    for e in pl.parallel(E):
        base = e * R
        n_rows = pl.read(recv_count, [e, 0])
        n_tiles = (n_rows + ROW_TILE - 1) // ROW_TILE
        for t in pl.parallel(n_tiles):
            t0 = base + t * ROW_TILE
            with pl.spmd(N // (MGI * N_TILE), name_hint="gate_mm"):
                pl.set_cache_policy(w1, C1)
                nb = pl.tile.get_block_idx()
                nbase = nb * (MGI * N_TILE)
                for ng in pl.range(MGI):
                    n0 = nbase + ng * N_TILE
                    acc = pl.create_tensor([1, ROW_TILE, N_TILE], dtype=pl.INT32)
                    for k0 in pl.pipeline(0, D, K_TILE, stage=2):
                        xk = x_flat[t0 : t0 + ROW_TILE, k0 : k0 + K_TILE]
                        wk = w1[e : e + 1, n0 : n0 + N_TILE, k0 : k0 + K_TILE]
                        if k0 == 0:
                            acc = pl.matmul(xk, wk, b_trans=True, out_dtype=pl.INT32)
                        else:
                            acc = pl.matmul_acc(acc, xk, wk, b_trans=True)
                    o1[t0 : t0 + ROW_TILE, n0 : n0 + N_TILE] = pl.reshape(acc, [ROW_TILE, N_TILE])
            with pl.spmd(N // (MGI * N_TILE), name_hint="up_mm"):
                pl.set_cache_policy(w3, C3)
                ub = pl.tile.get_block_idx()
                ubase = ub * (MGI * N_TILE)
                for ug in pl.range(MGI):
                    u0 = ubase + ug * N_TILE
                    uacc = pl.create_tensor([1, ROW_TILE, N_TILE], dtype=pl.INT32)
                    for uk in pl.pipeline(0, D, K_TILE, stage=2):
                        xu = x_flat[t0 : t0 + ROW_TILE, uk : uk + K_TILE]
                        wu = w3[e : e + 1, u0 : u0 + N_TILE, uk : uk + K_TILE]
                        if uk == 0:
                            uacc = pl.matmul(xu, wu, b_trans=True, out_dtype=pl.INT32)
                        else:
                            uacc = pl.matmul_acc(uacc, xu, wu, b_trans=True)
                    o3[t0 : t0 + ROW_TILE, u0 : u0 + N_TILE] = pl.reshape(uacc, [ROW_TILE, N_TILE])
            with pl.spmd(N // (MGI * N_TILE), name_hint="w2_mm"):
                pl.set_cache_policy(w2, C2)
                db = pl.tile.get_block_idx()
                dbase = db * (MGI * N_TILE)
                for dg in pl.range(MGI):
                    d0 = dbase + dg * N_TILE
                    dacc = pl.create_tensor([1, ROW_TILE, N_TILE], dtype=pl.INT32)
                    for dk in pl.pipeline(0, D, K_TILE, stage=2):
                        xd = x_flat[t0 : t0 + ROW_TILE, dk : dk + K_TILE]
                        wd = w2[e : e + 1, d0 : d0 + N_TILE, dk : dk + K_TILE]
                        if dk == 0:
                            dacc = pl.matmul(xd, wd, b_trans=True, out_dtype=pl.INT32)
                        else:
                            dacc = pl.matmul_acc(dacc, xd, wd, b_trans=True)
                    o2[t0 : t0 + ROW_TILE, d0 : d0 + N_TILE] = pl.reshape(dacc, [ROW_TILE, N_TILE])
    return o1, o3, o2


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--platform", default="a2a3")
    parser.add_argument("--device", type=int, default=0)
    args = parser.parse_args()

    torch.manual_seed(0)
    recv_x = torch.randint(-8, 8, (E, R, D), dtype=torch.int8)
    recv_count = torch.zeros(E, 1, dtype=torch.int32)
    recv_count[1, 0] = ROW_TILE
    w1 = torch.randint(-8, 8, (E, N, D), dtype=torch.int8)
    w3 = torch.randint(-8, 8, (E, N, D), dtype=torch.int8)
    w2 = torch.randint(-8, 8, (E, N, D), dtype=torch.int8)
    o1 = torch.zeros(E * R, N, dtype=torch.int32)
    o3 = torch.zeros(E * R, N, dtype=torch.int32)
    o2 = torch.zeros(E * R, N, dtype=torch.int32)

    cfg = RunConfig(platform=args.platform, device_id=args.device)
    compiled = three_mm.compile(config=cfg)
    compiled(recv_x, recv_count, w1, w3, w2, o1, o3, o2, config=cfg)

    t0 = 1 * R
    x = recv_x[1, 0:ROW_TILE].to(torch.int32)
    flags = f"b1={os.environ.get('L2G_B1','0')} b3={os.environ.get('L2G_B3','0')} b2={os.environ.get('L2G_B2','0')}"
    print(f"[aic3] {flags}")
    for name, o, w in (("o1/w1", o1, w1[1]), ("o3/w3", o3, w3[1]), ("o2/w2", o2, w2[1])):
        got = o[t0 : t0 + ROW_TILE]
        ref = x @ w.to(torch.int32).T
        bad = int((got != ref).sum())
        print(f"[aic3]   {name}: exact={torch.equal(got, ref)} mismatched={bad}/{ref.numel()} "
              f"max_abs_diff={int((got - ref).abs().max())}")
        if bad:
            # Did the aliased load land on some other live allocation?
            for cand_name, cand in (("w1[1]", w1[1]), ("w3[1]", w3[1]), ("w2[1]", w2[1]),
                                    ("w2[0]", w2[0]), ("w2[2]", w2[2]), ("w2[3]", w2[3]),
                                    ("zeros", torch.zeros_like(w2[1]))):
                if torch.equal(got, x @ cand.to(torch.int32).T):
                    print(f"[aic3]     -> bypassed load actually returned {cand_name}")
                    break
            else:
                print("[aic3]     -> does not match any other live tensor (garbage)")
```

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `c24df5e1` (branch: `main`; the repro is standalone) |
| pypto | `47b0ce53` (branch: `main`) |
| simpler | `09773588` (detached) |
| ptoas | `0.61` |
| pto-isa | `03e45c4b` (detached) |
| CANN | `9.0.0` |

All pins are consistent: simpler matches pypto's `runtime` gitlink, pto-isa
matches `runtime/pto_isa.pin`, ptoas matches `PTOAS_VERSION=v0.61`.

Diagnosis: **unclear** between PTOAS, pto-isa and simpler. Filed here because
PTOAS chose the lowering; see "Who fills the offset?" below. Happy to move it.

### Details

Since PTOAS 0.61 (#1356), `{cache_policy = #pto.load_cache_policy<l2_bypass>}`
on `pto.tload` lowers to:

```cpp
TLOAD<pto::TLoadL2Hint::NotAllocKeep>(v79, v88);
```

On a2a3 that template does **not** encode an address alias itself. pto-isa
(`include/pto/npu/a2a3/l2_cache_hint.hpp`) resolves it at run time:

```cpp
template <TLoadL2Hint l2Control, typename T>
PTO_INTERNAL T* ApplyTLoadL2HintAddr(T* addr)
{
    if constexpr (IsNotAllocTLoadL2Hint(l2Control)) {
        // Runtime-filled only (AscendC-compatible). CA/sim leaves 0 => addr+0.
        // NPU fills 0x100000000000 for real not-alloc. No hardcoded fallback.
        return reinterpret_cast<T*>(reinterpret_cast<uintptr_t>(addr) + g_opL2CacheHintCfg.l2Cacheoffset);
    }
    ...
}
```

`g_opL2CacheHintCfg` is an `inline __gm__` global emitted into each kernel
object, together with an `.ascend.meta` feature record (`{4, 4, 3}` = L2CACHE),
mirroring what AscendC emits so that a loader can fill it.

#### Who fills the offset?

Nothing in this stack provably does. `grep -rn "ascend.meta\|l2Cacheoffset\|g_opL2CacheHintCfg\|L2CACHE"`
over simpler's `src/` and `simpler_setup/` returns a single hit, an unrelated
comment in the a5 SIMT anchor header. simpler loads the incore `.bin` itself.

The value read on device is demonstrably **not 0**: with an offset of 0 the
aliased address equals the ordinary one, and the non-bypassed control is exact.
So the kernel adds a non-zero, unowned value to a valid GM address.

#### Measured behaviour

Three INT8 matmul stages (three AIC kernels, mirroring the model), one weight
each, bypass switched per stage. One free a2a3 card, `--device auto`:

| Config | Result |
|---|---|
| no bypass (control) | all three outputs bit-exact |
| bypass the **3rd** kernel's weight | **runs, wrong data** |
| bypass the **1st** kernel's weight | AICore exception `0x800000`, then `SCHEDULER_TIMEOUT S1:running-stalled` |

The corruption is confined to the bypassed load, and it is deterministic
(reproduced 3/3):

```text
[aic3] b1=0 b3=0 b2=1
[aic3]   o1/w1: exact=True  mismatched=0/32768      max_abs_diff=0
[aic3]   o3/w3: exact=True  mismatched=0/32768      max_abs_diff=0
[aic3]   o2/w2: exact=False mismatched=32761/32768  max_abs_diff=8247
[aic3]     -> does not match any other live tensor (garbage)
```

The real kernel fails the same way in every bypass configuration (w1 only, w2
only, all three), with golden FAIL or AICore exception; without bypass it passes.

#### The previously validated mechanism was self-contained

pypto-lib#1039 measured this optimisation by hand-editing generated C++, using a
**compile-time** alias on the weight base pointer:

```cpp
constexpr uint64_t L2_CACHE_DISABLE_OFFSET = 0x80000000000ULL;   // 2^43, A2/A3
__gm__ int8_t* weightL2Bypass = reinterpret_cast<__gm__ int8_t*>(
    reinterpret_cast<uint64_t>(v2) + L2_CACHE_DISABLE_OFFSET);
```

That form ran correctly on the same kernel (three AIC kernels, golden PASS) and
delivered the measured win. Note it is 2^43, while the pto-isa comment names
2^44 as the runtime-filled value; we have not verified which is correct for a2a3.

### What we would like

Either of these would make the feature usable; PTOAS is the layer that picks the
lowering, so the choice seems to belong here:

- lower `l2_bypass` to a self-contained address alias, so the emitted kernel does
  not depend on a runtime config nobody fills; or
- keep `TLoadL2Hint` and state the runtime contract explicitly (who fills
  `g_opL2CacheHintCfg`, with what value, and how a loader is expected to discover
  it), so simpler can implement it.

Independently: reading an unfilled config and silently corrupting results is the
worst failure mode. A zero/absent config should degrade to an ordinary cached
load, not to `addr + garbage`.

### Additional context

Related: hw-native-sys/PTOAS#1356 (the feature this lowering implements),
hw-native-sys/pypto#2680, hw-native-sys/pypto#2747, hw-native-sys/pypto-lib#1039.

We can run any diagnostic that helps, e.g. reading `g_opL2CacheHintCfg.l2Cacheoffset`
from inside a kernel and reporting the value.
